added link to JSON data of applications on applications/index.html.haml

hisayohorie opened the pull request:
I'm working on the issue #947, to create a link to see the list of applications as JSON data.
I haven't write the test yet as I have a question.

in this PR, I created application_json_path(authority) to direct user to /authorities/authority.short_name_encoded/application.js?key=current_user.api_key

I'm wondering if this link should *only* be on the applications index for a specific authority, or if it should also be on the general applications index like 
/applications.js?key=current_user.api_key  
(in which case the user needs bulk api access)

equivalentideas commented:
> I'm wondering if this link should only be on the applications index for a specific authority, or if it should also be on the general applications index like

I think for now just complete the issue (just the authority pages) and then we can build on that.

equivalentideas commented on app/helpers/applications_helper.rb:
```diff
@@ -68,6 +70,10 @@ def page_title(application)
     "#{@application.address} | #{@application.date_scraped.to_date.to_formatted_s(:rfc822)}"
   end
 
+  def applications_json_path(authority)
```
I think it would be good to name this something more descriptive, even if it's more verbose, like `authority_applications_json_path_for_current_user(authority)` so it's super clear what you're gonna get in the view.

equivalentideas replied:
That also matches the similar existing `authority_applications_path`.

equivalentideas commented on spec/helpers/applications_helper_spec.rb:
```diff
@@ -171,6 +171,18 @@
       end
     end
 
+    describe "applications_json_path" do
```
This does the job, but it makes the [separate test phases hard](https://robots.thoughtbot.com/four-phase-test) to read.

Also using [BetterSpecs](http://betterspecs.org/#subject) advice about using 'subject' etc. something like this might be clearer:

```
    describe "applications_json_path" do
      let(:authority) { create(:authority, short_name: "marrickville", ) }
      let(:user) { build(:user, api_key: "ABCDE12345") }

      before { expect(helper).to receive(:current_user).and_return(user) }

      subject { helper.applications_json_path(authority) }

      it { is_expected.to eq("/authorities/marrickville/applications.js?key=ABCDE12345") }
    end
```

Note that it uses `build` not `create` for the user, to avoid the before_create callback.

equivalentideas replied:
This should also point to the api subdomain, see https://github.com/openaustralia/planningalerts/blob/master/app/helpers/api_howto_helper.rb#L39-L42

and 

https://github.com/openaustralia/planningalerts/blob/master/app/helpers/api_howto_helper.rb#L11-L13

I think the best way to do that for now is to hard code it into a method on applications_helper like `api_host` to match the example. It should really be extracted into the theme, but if you don't feel like doing that now, I wouldn't worry about it. But implementing it the same way as in [`api_howto_helper.rb`](https://github.com/openaustralia/planningalerts/blob/master/app/helpers/api_howto_helper.rb#L11-L13) will help us extract it. If you could also add a comment like `# TODO: extract to theme` in the code, that would help.

equivalentideas commented on app/views/applications/index.html.haml:
```diff
@@ -14,5 +14,8 @@
     = link_to "Why?", faq_path(:anchor => "broken_scraper")
 
+- if current_user && @authority
+  = link_to "View as JSON", applications_json_path(@authority)
```
I think this would go better above the pagination links at the end of the list. This is because this is a kinda nice to have and we don't know if people will really use it—so better to not add more links to the already overcrowded top of the page. I think put it down there and api users will discover it. We can also add click tracking to it to see if people do.

hisayohorie commented on app/helpers/applications_helper.rb:
```diff
@@ -1,4 +1,6 @@
 module ApplicationsHelper
+
+
```
ah got it. will be mindful next time

hisayohorie commented on app/mailers/confirmation_mailer.rb:
```diff
@@ -1,6 +1,8 @@
 class ConfirmationMailer < ActionMailer::Base
   include ActionMailerThemer
 
+  add_template_helper(CommentsHelper)
```
got it!

equivalentideas commented:
Hey @hisayohorie looks like you've made progress 🚀 Let me know when I can review and merge this 🚢

hisayohorie commented:
@equivalentideas yes! I have changed:
- the name of authority_json_path to authority_applications_json_url_for_current_user
-  added api_host method in applications_helper.rb
- reformatted the spec

I think it is ready for your review!

equivalentideas commented:
Real nice work @hisayohorie 🚀 This has been merged 🔨 

I added a few things https://github.com/openaustralia/planningalerts/compare/fedb4bb27192...8c2f4965c71b (there's also a stray commit in there with some unrelated tests I wrote). You can read about them in the commit messages.

This is cool and it gave me a little idea for another way to point people to sign up for the API, ones we make some of the changes to commercial API sign up stuff ;-)

henare commented:
Is this deployed? I was excited to see it in action but I can't find it :eye:

equivalentideas commented:
> Is this deployed?

Nup, it was pretty late night here when I merged so set a reminder to deploy this morning. Should have noted that here 👍

henare commented:
>Nup, it was pretty late night here when I merged so set a reminder to deploy this morning.

No worries - good call :smile:

equivalentideas commented:
Deployed! https://www.planningalerts.org.au/authorities/act/applications nice one @hisayohorie :shipit:
